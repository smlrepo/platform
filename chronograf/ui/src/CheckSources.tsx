// Libraries
import React, {ReactElement, PureComponent} from 'react'
import {withRouter, InjectedRouter} from 'react-router'
import {connect} from 'react-redux'

// APIs
import {getSourceHealth} from 'src/sources/apis/v2'
import {getSourcesAsync} from 'src/shared/actions/sources'

// Actions
import {notify as notifyAction} from 'src/shared/actions/notifications'

// Constants
import * as copy from 'src/shared/copy/notifications'

// Types
import {Source} from 'src/types/v2'
import {Location} from 'history'
import {Notification, NotificationFunc, RemoteDataState} from 'src/types'
import {Links} from 'src/types/v2/links'

import {ErrorHandling} from 'src/shared/decorators/errors'

interface State {
  loading: RemoteDataState
}

interface Params {
  sourceID: string
}

interface Props {
  links: Links
  getSources: typeof getSourcesAsync
  sources: Source[]
  children: ReactElement<any>
  params: Params
  router: InjectedRouter
  location: Location
  notify: (message: Notification | NotificationFunc) => void
}

export const SourceContext = React.createContext()
// Acts as a 'router middleware'. The main `App` component is responsible for
// getting the list of data sources, but not every page requires them to function.
// Routes that do require data sources can be nested under this component.
@ErrorHandling
export class CheckSources extends PureComponent<Props, State> {
  constructor(props) {
    super(props)

    this.state = {
      loading: RemoteDataState.NotStarted,
    }
  }

  public async componentDidMount() {
    await this.props.getSources()
    this.setState({loading: RemoteDataState.Done})
  }

  public async componentDidUpdate() {
    const {loading} = this.state
    const {router, sources, notify} = this.props
    const source = this.source
    const defaultSource = sources.find(s => s.default === true)

    const isDoneLoading = loading === RemoteDataState.Done

    if (this.isRoot) {
      return router.push(`${this.rootPath}`)
    }

    if (isDoneLoading && !source) {
      if (defaultSource) {
        return router.push(`${this.path}?sourceID=${defaultSource.id}`)
      }

      if (sources[0]) {
        return router.push(`${this.path}?sourceID=${sources[0].id}`)
      }

      return router.push(`/sources/new?redirectPath=${this.path}`)
    }

    if (isDoneLoading) {
      try {
        await getSourceHealth(source.links.health)
      } catch (error) {
        notify(copy.sourceNoLongerAvailable(source.name))
      }
    }
  }

  public render() {
    const source = this.source
    if (this.isLoading || !source) {
      return <div className="page-spinner" />
    }

    return (
      <SourceContext.Provider value={source}>
        {this.props.children &&
          React.cloneElement(this.props.children, {...this.props, source})}
      </SourceContext.Provider>
    )
  }

  private get path(): string {
    const {location} = this.props

    if (this.isRoot) {
      return this.rootPath
    }

    return `${location.pathname}`
  }

  private get rootPath(): string {
    const {links, location} = this.props
    if (links.defaultDashboard) {
      const split = links.defaultDashboard.split('/')
      const id = split[split.length - 1]
      return `/dashboards/${id}${location.search}`
    }

    return `/dashboards`
  }

  private get isRoot(): boolean {
    const {
      location: {pathname},
    } = this.props

    return pathname === '' || pathname === '/'
  }

  private get isLoading(): boolean {
    const {loading} = this.state
    return (
      loading === RemoteDataState.Loading ||
      loading === RemoteDataState.NotStarted
    )
  }

  private get source(): Source {
    const {location, sources} = this.props
    return sources.find(s => s.id === location.query.sourceID)
  }
}

const mstp = ({sources, links}) => ({
  links,
  sources,
})

const mdtp = {
  getSources: getSourcesAsync,
  notify: notifyAction,
}

export default connect(mstp, mdtp)(withRouter(CheckSources))
