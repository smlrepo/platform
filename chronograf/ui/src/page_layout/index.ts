import Nav from 'src/page_layout/containers/Nav'
import Page from 'src/page_layout/components/Page'
import PageHeader from 'src/page_layout/components/PageHeader'
import PageContents from 'src/page_layout/components/PageContents'

export {Page, PageHeader, PageContents}
export default Nav
