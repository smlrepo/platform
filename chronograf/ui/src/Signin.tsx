// Libraries
import React, {ReactElement, PureComponent} from 'react'
import {connect} from 'react-redux'

// APIs
import {trySources} from 'src/onboarding/apis'

// Components
import {ErrorHandling} from 'src/shared/decorators/errors'
import SigninPage from 'src/onboarding/containers/SigninPage'
import Notifications from 'src/shared/components/notifications/Notifications'

// Types
import {RemoteDataState} from 'src/types'
import {Links} from 'src/types/v2/links'

interface State {
  loading: RemoteDataState
  isSourcesAllowed: boolean
}

interface Props {
  links: Links
  children: ReactElement<any>
}

@ErrorHandling
export class Signin extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      loading: RemoteDataState.NotStarted,
      isSourcesAllowed: false,
    }
  }

  public async componentDidMount() {
    const {links} = this.props
    const isSourcesAllowed = await trySources(links.sources)
    this.setState({loading: RemoteDataState.Done, isSourcesAllowed})
  }

  public render() {
    const {isSourcesAllowed} = this.state

    if (this.isLoading) {
      return <div className="page-spinner" />
    }
    if (!isSourcesAllowed) {
      return (
        <div className="chronograf-root">
          <Notifications inPresentationMode={true} />
          <SigninPage />
        </div>
      )
    } else {
      return this.props.children && React.cloneElement(this.props.children)
    }
  }

  private get isLoading(): boolean {
    const {loading} = this.state
    return (
      loading === RemoteDataState.Loading ||
      loading === RemoteDataState.NotStarted
    )
  }
}

const mstp = ({links}) => ({
  links,
})

export default connect(mstp)(Signin)
