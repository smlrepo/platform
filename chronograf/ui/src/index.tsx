import 'babel-polyfill'

import React, {PureComponent} from 'react'
import {render} from 'react-dom'
import {Provider} from 'react-redux'
import {Router, Route, useRouterHistory} from 'react-router'
import {createHistory} from 'history'
import {syncHistoryWithStore} from 'react-router-redux'
import {bindActionCreators} from 'redux'

import configureStore from 'src/store/configureStore'
import {loadLocalStorage} from 'src/localStorage'

import {getRootNode} from 'src/utils/nodes'
import {getBasepath} from 'src/utils/basepath'

import App from 'src/App'
import CheckSources from 'src/CheckSources'
import {DashboardsPage, DashboardPage} from 'src/dashboards'
import {SourcePage, ManageSources} from 'src/sources'
import {FluxPage} from 'src/flux'
import NotFound from 'src/shared/components/NotFound'

import {getLinksAsync} from 'src/shared/actions/links'

import {disablePresentationMode} from 'src/shared/actions/app'
import {errorThrown} from 'src/shared/actions/errors'
import {notify} from 'src/shared/actions/notifications'

import 'src/style/chronograf.scss'

import * as ErrorsModels from 'src/types/errors'

const errorsQueue = []

const rootNode = getRootNode()

const basepath = getBasepath()

declare global {
  interface Window {
    basepath: string
  }
}

// Older method used for pre-IE 11 compatibility
window.basepath = basepath

const browserHistory = useRouterHistory(createHistory)({
  basename: basepath, // this is written in when available by the URL prefixer middleware
})

const store = configureStore(loadLocalStorage(errorsQueue), browserHistory)
const {dispatch} = store

browserHistory.listen(() => {
  dispatch(disablePresentationMode())
})

window.addEventListener('keyup', event => {
  const escapeKeyCode = 27
  // fallback for browsers that don't support event.key
  if (event.key === 'Escape' || event.keyCode === escapeKeyCode) {
    dispatch(disablePresentationMode())
  }
})

const history = syncHistoryWithStore(browserHistory, store)

interface State {
  ready: boolean
}

class Root extends PureComponent<{}, State> {
  private getLinks = bindActionCreators(getLinksAsync, dispatch)

  constructor(props) {
    super(props)
    this.state = {
      ready: false,
    }
  }

  public async componentDidMount() {
    this.flushErrorsQueue()

    try {
      await this.getLinks()
      this.setState({ready: true})
    } catch (error) {
      dispatch(errorThrown(error))
    }
  }

  public render() {
    return this.state.ready ? (
      <Provider store={store}>
        <Router history={history}>
          <Route component={App}>
            <Route path="/" component={CheckSources}>
              <Route path="dashboards/:dashboardID" component={DashboardPage} />
              <Route path="sources/new" component={SourcePage} />
              <Route path="dashboards" component={DashboardsPage} />
              <Route path="manage-sources" component={ManageSources} />
              <Route path="manage-sources/new" component={SourcePage} />
              <Route path="manage-sources/:id/edit" component={SourcePage} />
              <Route path="delorean" component={FluxPage} />
            </Route>
          </Route>
          <Route path="*" component={NotFound} />
        </Router>
      </Provider>
    ) : (
      <div className="page-spinner" />
    )
  }

  private flushErrorsQueue() {
    if (errorsQueue.length) {
      errorsQueue.forEach(error => {
        if (typeof error === 'object') {
          dispatch(notify(error))
        } else {
          dispatch(
            errorThrown(
              {status: 0, auth: null},
              error,
              ErrorsModels.AlertType.Warning
            )
          )
        }
      })
    }
  }
}

if (rootNode) {
  render(<Root />, rootNode)
}
