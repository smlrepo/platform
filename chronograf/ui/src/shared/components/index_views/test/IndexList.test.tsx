import React from 'react'
import {mount} from 'enzyme'

import {Alignment} from 'src/clockface'
import IndexList from 'src/shared/components/index_views/IndexList'

describe('IndexList', () => {
  let wrapper

  const wrapperSetup = (override = {}) => {
    const emptyState = <div>Empty</div>

    const columns = [
      {
        key: 'test--fruit',
        size: 500,
        title: 'Fruit',
        align: Alignment.Left,
        showOnHover: false,
      },
      {
        key: 'test--calories',
        size: 500,
        title: 'Calories',
        align: Alignment.Left,
        showOnHover: false,
      },
    ]

    const rows = [
      {
        disabled: false,
        columns: [
          {
            key: 'test--fruit',
            contents: 'Apple',
          },
          {
            key: 'test--calories',
            contents: '500',
          },
        ],
      },
      {
        disabled: false,
        columns: [
          {
            key: 'test--fruit',
            contents: 'Pear',
          },
          {
            key: 'test--calories',
            contents: '1000',
          },
        ],
      },
      {
        disabled: false,
        columns: [
          {
            key: 'test--fruit',
            contents: 'Banana',
          },
          {
            key: 'test--calories',
            contents: '200',
          },
        ],
      },
    ]

    const props = {
      columns,
      rows,
      emptyState,
      ...override,
    }

    return mount(<IndexList {...props} />)
  }

  it('mounts without exploding', () => {
    wrapper = wrapperSetup()
    expect(wrapper).toHaveLength(1)
  })

  it('matches snapshot with minimal props', () => {
    wrapper = wrapperSetup()
    expect(wrapper).toMatchSnapshot()
  })

  it('renders empty state when 0 rows exist', () => {
    wrapper = wrapperSetup({rows: []})

    const emptyDiv = wrapper
      .find('div')
      .filterWhere(div => div.prop('data-test'))

    expect(emptyDiv.prop('data-test')).toBe('empty-state')
  })

  it('matches snapshot when 0 rows exist', () => {
    wrapper = wrapperSetup({rows: []})
    expect(wrapper).toMatchSnapshot()
  })
})
