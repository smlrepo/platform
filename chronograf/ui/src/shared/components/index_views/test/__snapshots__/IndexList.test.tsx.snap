// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexList matches snapshot when 0 rows exist 1`] = `
<IndexList
  columns={
    Array [
      Object {
        "align": "left",
        "key": "test--fruit",
        "showOnHover": false,
        "size": 500,
        "title": "Fruit",
      },
      Object {
        "align": "left",
        "key": "test--calories",
        "showOnHover": false,
        "size": 500,
        "title": "Calories",
      },
    ]
  }
  emptyState={
    <div>
      Empty
    </div>
  }
  rows={Array []}
>
  <table
    className="index-list"
  >
    <IndexListHeader
      columns={
        Array [
          Object {
            "align": "left",
            "key": "test--fruit",
            "showOnHover": false,
            "size": 500,
            "title": "Fruit",
          },
          Object {
            "align": "left",
            "key": "test--calories",
            "showOnHover": false,
            "size": 500,
            "title": "Calories",
          },
        ]
      }
      getColumnWidthPercent={[Function]}
    >
      <thead
        className="index-list--header"
      >
        <tr>
          <th
            className="index-list--header-cell index-list--align-left"
            key="test--fruit"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            Fruit
          </th>
          <th
            className="index-list--header-cell index-list--align-left"
            key="test--calories"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            Calories
          </th>
        </tr>
      </thead>
    </IndexListHeader>
    <tbody
      className="index-list--empty"
    >
      <tr
        className="index-list--empty-row"
      >
        <td
          colSpan={2}
        >
          <div
            className="index-list--empty-cell"
            data-test="empty-state"
          >
            <div>
              Empty
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</IndexList>
`;

exports[`IndexList matches snapshot with minimal props 1`] = `
<IndexList
  columns={
    Array [
      Object {
        "align": "left",
        "key": "test--fruit",
        "showOnHover": false,
        "size": 500,
        "title": "Fruit",
      },
      Object {
        "align": "left",
        "key": "test--calories",
        "showOnHover": false,
        "size": 500,
        "title": "Calories",
      },
    ]
  }
  emptyState={
    <div>
      Empty
    </div>
  }
  rows={
    Array [
      Object {
        "columns": Array [
          Object {
            "contents": "Apple",
            "key": "test--fruit",
          },
          Object {
            "contents": "500",
            "key": "test--calories",
          },
        ],
        "disabled": false,
      },
      Object {
        "columns": Array [
          Object {
            "contents": "Pear",
            "key": "test--fruit",
          },
          Object {
            "contents": "1000",
            "key": "test--calories",
          },
        ],
        "disabled": false,
      },
      Object {
        "columns": Array [
          Object {
            "contents": "Banana",
            "key": "test--fruit",
          },
          Object {
            "contents": "200",
            "key": "test--calories",
          },
        ],
        "disabled": false,
      },
    ]
  }
>
  <table
    className="index-list"
  >
    <IndexListHeader
      columns={
        Array [
          Object {
            "align": "left",
            "key": "test--fruit",
            "showOnHover": false,
            "size": 500,
            "title": "Fruit",
          },
          Object {
            "align": "left",
            "key": "test--calories",
            "showOnHover": false,
            "size": 500,
            "title": "Calories",
          },
        ]
      }
      getColumnWidthPercent={[Function]}
    >
      <thead
        className="index-list--header"
      >
        <tr>
          <th
            className="index-list--header-cell index-list--align-left"
            key="test--fruit"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            Fruit
          </th>
          <th
            className="index-list--header-cell index-list--align-left"
            key="test--calories"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            Calories
          </th>
        </tr>
      </thead>
    </IndexListHeader>
    <tbody
      className="index-list--body"
    >
      <IndexListRow
        disabled={false}
        getColumnWidthPercent={[Function]}
        getRowColumnClassName={[Function]}
        key="index-list--row-0"
        rowColumns={
          Array [
            Object {
              "contents": "Apple",
              "key": "test--fruit",
            },
            Object {
              "contents": "500",
              "key": "test--calories",
            },
          ]
        }
        rowIndex={0}
      >
        <tr
          className="index-list--row"
        >
          <td
            className="index-list--row-cell index-list--align-left"
            key="index-list--row-0-col-test--fruit"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <div
              className="index-list--cell"
            >
              Apple
            </div>
          </td>
          <td
            className="index-list--row-cell index-list--align-left"
            key="index-list--row-0-col-test--calories"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <div
              className="index-list--cell"
            >
              500
            </div>
          </td>
        </tr>
      </IndexListRow>
      <IndexListRow
        disabled={false}
        getColumnWidthPercent={[Function]}
        getRowColumnClassName={[Function]}
        key="index-list--row-1"
        rowColumns={
          Array [
            Object {
              "contents": "Pear",
              "key": "test--fruit",
            },
            Object {
              "contents": "1000",
              "key": "test--calories",
            },
          ]
        }
        rowIndex={1}
      >
        <tr
          className="index-list--row"
        >
          <td
            className="index-list--row-cell index-list--align-left"
            key="index-list--row-1-col-test--fruit"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <div
              className="index-list--cell"
            >
              Pear
            </div>
          </td>
          <td
            className="index-list--row-cell index-list--align-left"
            key="index-list--row-1-col-test--calories"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <div
              className="index-list--cell"
            >
              1000
            </div>
          </td>
        </tr>
      </IndexListRow>
      <IndexListRow
        disabled={false}
        getColumnWidthPercent={[Function]}
        getRowColumnClassName={[Function]}
        key="index-list--row-2"
        rowColumns={
          Array [
            Object {
              "contents": "Banana",
              "key": "test--fruit",
            },
            Object {
              "contents": "200",
              "key": "test--calories",
            },
          ]
        }
        rowIndex={2}
      >
        <tr
          className="index-list--row"
        >
          <td
            className="index-list--row-cell index-list--align-left"
            key="index-list--row-2-col-test--fruit"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <div
              className="index-list--cell"
            >
              Banana
            </div>
          </td>
          <td
            className="index-list--row-cell index-list--align-left"
            key="index-list--row-2-col-test--calories"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <div
              className="index-list--cell"
            >
              200
            </div>
          </td>
        </tr>
      </IndexListRow>
    </tbody>
  </table>
</IndexList>
`;
